pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -225,11 +235,81 @@ func getServiceEntryHandler(c *ServiceEntryStore) func(model.Config, model.Confi
 	}
 }
 
+// GetForeignServiceInstanceHandler defines the handler for service instances generated by other registries
+func (d *ServiceEntryStore) GetForeignServiceInstanceHandler() func(*model.ServiceInstance, model.Event) {
+	return func(si *model.ServiceInstance, event model.Event) {
+		key := configKey{
+			kind:      foreignInstanceConfigType,
+			name:      si.Endpoint.Address,
+			namespace: si.Service.Attributes.Namespace,
+		}
+		// Used to indicate if this event was fired for a pod->workloadentry conversion
+		// and that the event can be ignored due to no relevant change in the workloadentry
+		redundantEventForPod := false
+
+		d.storeMutex.Lock()
+		// We will only select entries in the same namespace
+		entries := d.seWithSelectorByNamespace[si.Service.Attributes.Namespace]
+
+		// this is from a pod. Store it in separate map so that
+		// the refreshIndexes function can use these as well as the store ones.
+		switch event {
+		case model.EventDelete:
+			if _, exists := d.foreignRegistryInstancesByIP[si.Endpoint.Address]; !exists {
+				// multiple delete events for the same pod (succeeded/failed/unknown status repeating).
+				redundantEventForPod = true
+			} else {
+				delete(d.foreignRegistryInstancesByIP, si.Endpoint.Address)
+			}
+		default: // add or update
+			if old, exists := d.foreignRegistryInstancesByIP[si.Endpoint.Address]; exists {
+				// If multiple k8s services select the same pod, we may be getting multiple events
+				// ignore them as we only care about the Endpoint itself.
+				if reflect.DeepEqual(old.Endpoint, si.Endpoint) {
+					// ignore the udpate as nothing has changed
+					redundantEventForPod = true
+				}
+			}
+			d.foreignRegistryInstancesByIP[si.Endpoint.Address] = si
```

Comment:
I am worried about the memory usage

Reply:
We should only cache those we care about, those that can be selected by SE

Reply:
Agreed, and also before it is merged into 1.6 - make sure a flag is added !

Don't know how I forgot about it, but we shouldn't add features in last minute on by default and without a flag.